servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/NettyHttpServer.java

```diff
@@ -420,22 +422,36 @@ public void onComplete() {
                         }));
 
                 if (drainRequestPayloadBody) {
-                    responseWrite = responseWrite.concat(defer(() -> (payloadSubscribed.get() ?
-                            // Discarding the request payload body is an operation which should not impact the state of
-                            // request/response processing. It's appropriate to recover from any error here.
-                            // ST may introduce RejectedSubscribeError if user already consumed the request payload body
-                            requestCompletion : request.messageBody().ignoreElements().onErrorComplete())
-                            // No need to make a copy of the context in both cases.
-                            .shareContextOnSubscribe())
-                            // We need to apply shareContextOnSubscribe() on deferred Completable to share the same
-                            // context between concatenated Completables.
-                            .shareContextOnSubscribe());
-                } else {
-                    responseWrite = responseWrite.concat(requestCompletion);
+                    responseWrite = responseWrite.afterFinally(new TerminalSignalConsumer() {
+                        @Override
+                        public void onComplete() {
+                            if (payloadSubscribed.compareAndSet(false, true)) {
+                                // At this point, discarding the request payload body is an operation which should not
+                                // impact the state of request/response processing, and we need to do it in order to
+                                // both reuse the connection and finish any operators that may depend on both the
+                                // request and response bodies to have been fully processed.
+                                request.messageBody().ignoreElements().shareContextOnSubscribe().subscribe();
```

Comment:
Do we also need to `shareContextOnSubscribe` if the subscriber doesn't do anything with it?

Reply:
In this it's a bit unclear since the user intent doesn't exist. However, we generally lean toward sharing if it's tasks that we initiate because users can always induce a copy via a `defer(..)` block but it's not possible to 'uncopy' if ST chose to make one on their behalf.

Reply:
Even though the subscriber doesn't do anything with it, it triggers callbacks that users and ServiceTalk attached to the request payload. If you remove `shareContextOnSubscribe()` from here and run all `AbstractHttpServiceAsyncContextTest`s, you will see that `onRequest*` callbacks in `HttpLifecycleObserver` report:
```
Caused by: java.lang.AssertionError: 
Expected: is sameInstance(<CopyOnWriteContextMap@4639d211:{ContextMap.Key{name='k1', type=CharSequence}@2d1e6c90=1}>)
     but: was <CopyOnWriteContextMap@42065e0e:{ContextMap.Key{name='k1', type=CharSequence}@2d1e6c90=1}>
```
The intend is to propagate exact same instance of `AsyncContext` through all parts or request and response. Then if users need to isolate context themselves, they can do that with `defer` when necessary.